app/common/state/historyState.js

```diff
@@ -0,0 +1,174 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const {BrowserWindow} = require('electron')
+const tabState = require('./tabState.js')
+const config = require('../../../js/constants/config.js')
+const eventUtil = require('../../../js/lib/eventUtil.js')
+const appActions = require('../../../js/actions/appActions.js')
+const UrlUtil = require('../../../js/lib/urlutil.js')
+const messages = require('../../../js/constants/messages')
+const locale = require('../../locale.js')
+const CommonMenu = require('../commonMenu.js')
+
+const onBackButtonHistoryMenu = (tabId, partitionNumber, history, rect) => {
+  const menuTemplate = []
+
+  if (tabId > -1 && history && history.entries.length > 0) {
+    const stopIndex = Math.max(((history.currentIndex - config.navigationBar.maxHistorySites) - 1), -1)
+    for (let index = (history.currentIndex - 1); index > stopIndex; index--) {
+      const url = history.entries[index].url
+
+      menuTemplate.push({
+        label: history.entries[index].display,
+        icon: history.entries[index].icon,
+        click: (e) => {
+          if (eventUtil.isForSecondaryAction(e)) {
+            appActions.createTabRequested({
+              url,
+              partitionNumber: partitionNumber,
+              active: !!e.shiftKey
+            })
+          } else {
+            appActions.onNavigateIndex(tabId, index)
+          }
+        }
+      })
+    }
+
+    // Always display "Show History" link
+    menuTemplate.push(
+      CommonMenu.separatorMenuItem,
+      {
+        label: locale.translation('showAllHistory'),
+        click: () => {
+          appActions.createTabRequested({
+            url: 'about:history'
+          })
+          BrowserWindow.getFocusedWindow().webContents.send(messages.HIDE_CONTEXT_MENU)
+        }
+      })
+  }
+
+  if (BrowserWindow.getFocusedWindow()) {
+    BrowserWindow.getFocusedWindow().webContents.send(messages.SHOW_CONTEXT_MENU, {
```

Comment:
Our dispatcher now has the ability to send messages to a particular window.  So it's more preferable to do an app action with a window Id specified.  See for example `newWebContentsAdded`.